pkg/vault/contracts/AggregatorBatchRouter.sol

```diff
@@ -0,0 +1,476 @@
+                IERC20 stepTokenIn;
+
+                if (stepLocals.isFirstStep) {
+                    // The first step in the iteration is the last one in the given array of steps, and it
+                    // specifies the output token for the step as well as the exact amount out for that token.
+                    // Output amounts are stored to send them later on.
+                    _currentSwapTokensOut().add(address(step.tokenOut));
+                    _currentSwapTokenOutAmounts().tAdd(address(step.tokenOut), stepExactAmountOut);
+                }
+
+                if (stepLocals.isLastStep) {
+                    // In backwards order, the last step is the first one in the given path.
+                    // The given token in and max amount in apply for this step.
+                    stepMaxAmountIn = path.maxAmountIn;
+                    stepTokenIn = path.tokenIn;
+                } else {
+                    // For every other intermediate step, no maximum input applies.
+                    // The input token for this step is the output token of the previous given step.
+                    // We use uint128 to prevent Vault's internal scaling from overflowing.
+                    stepMaxAmountIn = _MAX_AMOUNT;
+                    stepTokenIn = path.steps[uint256(j - 1)].tokenOut;
+                }
+
+                uint256 amountIn;
+                if (step.isBuffer) {
+                    (, amountIn, ) = _vault.erc4626BufferWrapOrUnwrap(
+                        BufferWrapOrUnwrapParams({
+                            kind: SwapKind.EXACT_OUT,
+                            direction: step.pool == address(stepTokenIn)
+                                ? WrappingDirection.UNWRAP
+                                : WrappingDirection.WRAP,
+                            wrappedToken: IERC4626(step.pool),
+                            amountGivenRaw: stepExactAmountOut,
+                            limitRaw: stepMaxAmountIn
+                        })
+                    );
+                } else {
+                    // No BPT involved in the operation: regular swap exact out.
+                    (, amountIn, ) = _vault.swap(
+                        VaultSwapParams({
+                            kind: SwapKind.EXACT_OUT,
+                            pool: step.pool,
+                            tokenIn: stepTokenIn,
+                            tokenOut: step.tokenOut,
+                            amountGivenRaw: stepExactAmountOut,
+                            limitRaw: stepMaxAmountIn,
+                            userData: params.userData
+                        })
+                    );
+                }
+
+                if (stepLocals.isLastStep) {
+                    pathAmountsIn[i] = amountIn;
+                    _currentSwapTokensOut().add(address(stepTokenIn));
+                    _currentSwapTokenOutAmounts().tAdd(address(stepTokenIn), path.maxAmountIn - amountIn);
```

Comment:
I'd clarify that this is to reimburse the difference between max amount in and actual amount in, and we're just adding it to the token out amounts to process altogether at settlement time (which I think is a good idea).